curator-framework/src/main/java/org/apache/curator/framework/imps/Watching.java

```diff
@@ -85,10 +95,8 @@ void commitWatcher(int rc, boolean isExists) {
             doCommit = (rc == KeeperException.Code.OK.intValue());
         }
 
-        if (doCommit && (namespaceWatcher != null)) {
-            if (client.getWatcherRemovalManager() != null) {
-                client.getWatcherRemovalManager().add(namespaceWatcher);
-            }
+        if (doCommit && namespaceWatcher != null && watcherRemovalManager != null) {
```

Comment:
Can you elaborate a bit when `watcherRemovalManager` can be different from `client.getWatcherRemovalManager()`.

Reply:
I agree with the question: we can save some memory by not having the field. The performance impact is negligible

Reply:
1. Given a fresh new `CuratorFrameworkImpl`, say `framework0`.
2. Let's assume `framework1 = framework0.newWatcherRemoveCuratorFramework()`. `framework1.getWatcherRemovalManager` will be the `WatcherRemovalManager`.
3. Let's assume `framework2 = frame1.usingNamespace(null)`. `framework2` will loss `WatcherRemovalManager` from above.
5. `framework1.removeWatchers` will not drop watchers from `new Watching(framework2, ...)`.

Step.2 is what exactly `EnsembleTracker` does.
https://github.com/apache/curator/blob/6a27ff0a17df9ecee994ed06acfc2de6cbde58cf/curator-framework/src/main/java/org/apache/curator/framework/imps/EnsembleTracker.java#L87-L90

Step.3 is what exactly `GetConfigBuilderImpl`(CURATOR-667(#474)) does currently.
https://github.com/apache/curator/blob/6a27ff0a17df9ecee994ed06acfc2de6cbde58cf/curator-framework/src/main/java/org/apache/curator/framework/imps/GetConfigBuilderImpl.java#L41-L45

Step.4 is where this bug emerges.
https://github.com/apache/curator/blob/6a27ff0a17df9ecee994ed06acfc2de6cbde58cf/curator-framework/src/main/java/org/apache/curator/framework/imps/EnsembleTracker.java#L99-L104

I have pushed a fixup commit 6b78a3b849b0136d4a45225981c099e8350f4383 with comments to doc this.